### PR TITLE
Separate generation/training + distributed training

### PR DESCRIPTION
**Pull request by dcfidalgo**
So, this should be ready, finally ...
With these changes, we can separate the data generation and the segmentation model training, and distribute both over multiple nodes/GPUs. With the current scripts, generating 500,000 image-label pairs should take around ~5-6 hours (20 GPUs), and the respective training should also take 6h max (4 GPUs).

Quick summary of the main changes in each file:
- `brain_generator.py`: added functionality to write image-label pairs to TFRecords, reading them for training, as well as converting them to the output format of `BrainGenerator.generate_brain`;
- `metrics_model.py`: Copied the Loss functions to loss classes to be used with `keras.Model.compile`;
- `segmentation_model.py`: Dummy implementation of the segmentation UNet, i will provide a proper implementation in a follow-up PR. The original SynthSeg implementation has a memory lead when distributing it over multiple GPUs;
- `training.py`: Now returns results in the History object. Used to compare the new training loop with the original one;
- `training_options.py`: Added new options for the new training loop;
- `training_with_tfrecords.py`: Implementation of the new training loop using TFRecords created previously;
- `generation.py`: Add functionality of generating TFRecords instead of `nii.gz`;
- `generation.slurm`: Generation slurm script;
- `training.py`: Fixed stuff, add functionality to train with tfrecords
- `training.slurm`: Training slurm script;
- added some tests

@halirutan Maybe we could have a quick call this week to go over it together?

### CODE REVIEW

File: tests/test_training_with_tfrecords.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    ids=["wl2", "dice", "dice_after_wl2"],
+)
+def test_training(
+    tmp_path, wl2_epochs, dice_epochs, mean, std, exact, files
+):
+    """
+    Tests the equivalence with the original training via `mean` and `std`
+    and the current implementation via `exact`.
+    """
+    tf.keras.utils.set_random_seed(43)
```
Comment (halirutan) [inline tests/test_training_with_tfrecords.py:49]: Would it make sense to use our fixture for pinning random seed initialization here? We have tests.conftest.fixed_random_seed and I think we only need to add the tf.keras.utils.set_random_seed there.

Comment (halirutan) [reply tests/test_training_with_tfrecords.py:49]: On a closer look it seems that the keras utils function does exactly what we do in our fixture. Should we just replace all calls to our fixture with the keras helper?

File: SynthSeg/training_with_tfrecords.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        opts: The training options. The parameters related to the generation of the synthetic images will be ignored.
+    """
+    # Check epochs
+    assert (opts.wl2_epochs > 0) | (
+        opts.dice_epochs > 0
+    ), "either wl2_epochs or dice_epochs must be positive, had {0} and {1}".format(
+        opts.wl2_epochs, opts.dice_epochs
+    )
+
+    # Make sure ints are recognized as such
```
Comment (halirutan) [inline SynthSeg/training_with_tfrecords.py]: What happens if the user supplies "epochs" which is a valid value, according to the training_options.py docs?

File: scripts/slurm/training.slurm [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+#SBATCH --mem=0
+#
+#SBATCH --constraint="gpu"
+#SBATCH --gres=gpu:a100:4
+#
+#SBATCH --mail-type=none
+#SBATCH --mail-user=david.carreto.fidalgo@mpcdf.mpg.de
+#SBATCH --time=00:15:00
+
+cfg_file=/ptmp/dcfidalgo/projects/cbs/segmentation/experiments/reproduce/train.yml
```
Comment (halirutan) [inline scripts/slurm/training.slurm:22]: Why did we switch from .json to .yml files for the configuration?

File: scripts/slurm/generation.slurm [hunk 4]
```diff
@@ -10,25 +10,22 @@
...
 #SBATCH --gres=gpu:a100:1
 #
 #SBATCH --mail-type=none
 #SBATCH --mail-user=david.carreto.fidalgo@mpcdf.mpg.de
-#SBATCH --time=01:00:00
+#SBATCH --time=00:20:00
+
+
+output_dir=/ptmp/dcfidalgo/projects/cbs/segmentation/data/reproduce
+config_file=${output_dir}/config.yml
```
Comment (halirutan) [inline scripts/slurm/generation.slurm:22]: Same question about the .yml vs .json. My main issue is that we should be consistent. Does the .yml representation have considerable advantages over .json?

File: SynthSeg/metrics_model.py [hunk 5]
```diff
@@ -69,3 +69,75 @@ def loss(self, y_true, y_predicted):
 
         tf.debugging.check_numerics(loss, 'Loss not finite')
         return loss
+
+
+class WeightedL2Loss:
+    def __init__(
+        self, n_labels: int, target_value: float = 5, background_weight: float = 1e-4
```
Comment (halirutan) [inline SynthSeg/metrics_model.py]: 5.0 and 1.0e-4, right?

File: SynthSeg/segmentation_model.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
+from tensorflow import keras
+from typing import Tuple
+
+
+def unet(input_shape: Tuple[int, ...], num_classes):
```
Comment (halirutan) [inline SynthSeg/segmentation_model.py:5]: I think you mentioned that you wanted to re-implement the UNet. However, this doesn't seem to support all the training options like convolution size yet, right? What's the plan for that?

**Review by halirutan — APPROVED** [review]
This looks good to me. I have several questions regarding the meaning of some things and if you thought of some implications. But in general, I think this is good to go. Let's have a chat.

File: tests/test_training_with_tfrecords.py (hunk 1 above)
Comment (dcfidalgo) [reply tests/test_training_with_tfrecords.py:49]: Arg, sorry, wasn't aware of the fixture ... i think it seems reasonable to remove the fixture and use the keras helper instead. Will add a commit to this PR.

File: SynthSeg/training_with_tfrecords.py (hunk 2 above)
Comment (dcfidalgo) [reply SynthSeg/training_with_tfrecords.py]: Then the exception is caught and the value is still "epochs". But I just found out that the order in the type declaration matters to simple parsing, we can remove this try-except block.

File: scripts/slurm/training.slurm (hunk 3 above)
Comment (dcfidalgo) [reply scripts/slurm/training.slurm:22]: no particular reason, i just find yml files easier/quicker to write than json files so I generally prefer them for cfg stuff, e.g.
**YAML**:
```yaml
labels_dir: ../../data/training_label_maps
generation_labels: null
batchsize: 1
scaling_bounds: 0.2
output_shape: [1, 2, 3]
flipping: true
```
**JSON**
```json
{
  "labels_dir": "../../data/training_label_maps",
  "generation_labels": null,
  "batchsize": 1,
  "scaling_bounds": 0.2,
  "output_shape": [1, 2, 3],
  "flipping": true
}
```

File: SynthSeg/metrics_model.py (hunk 5 above)
Comment (dcfidalgo) [reply SynthSeg/metrics_model.py]: right! :+1:

File: SynthSeg/segmentation_model.py (hunk 6 above)
Comment (dcfidalgo) [reply SynthSeg/segmentation_model.py:5]: Yes, currently this is just a dummy to test that the distributed training works. I am working on a proper implementation, will be a follow-up PR to make this PR a bit lighter.
